### DISCUSSION TITLE
Editor de Textos - Do/While na prática

### DISCUSSION THREAD

**Discussion by Renato-Damasceno** [balta-io/2801#11] [Perguntas e Respostas] [answered]
Boa noite,

Gostaria de entender melhor por que o sistema está excluindo o primeiro caractere do texto?

```
            do
            {
                text += Console.ReadLine();
                text += Environment.NewLine;
                
            }
            while (Console.ReadKey().Key != ConsoleKey.Escape);
```

**Comment by andrebaltieri** [comment]
Bom dia @Renato-Damasceno poderia postar um print?

**Reply by mmoraesADS** [reply]
Também estou com o mesmo problema, a partir da segunda linha do editor de texto, fica comendo a primeira letra de cada linha.
![Captura de Tela 2021-09-08 às 21 55 33](https://user-images.githubusercontent.com/56050685/132604711-482963b5-1253-4436-8cff-99af0d21c600.png)

**Reply by RafaelDifini** [reply]
@andrebaltieri Estou com o mesmo problema no EditorHtml, no curso vc diz pra simplesmente dar um espaço porem nao funciona 
![image](https://user-images.githubusercontent.com/79666617/155240297-2711ac5f-bdf9-4a6d-aa8e-079815f30145.png)

**Comment by BrewertonSantos** [comment, marked as answer]
Quando a tecla Enter é pressionada o console entende como uma saída. Então executa a ação para ler o que foi escrito na entrada e logo em seguida executa a verificação de caractere através do Console.ReadKey(). Para isto a aplicação captura o primeiro caractere da entrada.

Ao concluir que a tecla inicial não é equivalente à apresentada na condição (Escape), ele continua a execução do código mas o caractere capturado é excluído do resultado. Por isto ele some da exibição a seguir. Existem diversas formas de burlar este comportamento. Ficando a critério do desenvolvedor a melhor solução para o caso de uso em questão.

**Reply by Renato-Damasceno** [reply]
Teria alguma dica de como posso resolver este problema?

**Reply by andrebaltieri** [reply]
Uma abordagem seria ler caractere por caractere e armazenar na variável `text`. Caso o último caractere lido seja `ESC` sai do programa.
Outra abordagem seria ter um loop infinito e dada uma palavra como `:wq` (Usada no VIM) o editor sai.

**Reply by andrebaltieri** [reply]
@RafaelDifini veja nesta thread aqui

**Reply by kfurfles** [reply]
Exemplo da abordagem citada acima

 ```csharp
 static void Editar() {
    Console.Clear();
    Console.WriteLine("Digite seu texto abaixo (ESC) para sair");
    Console.WriteLine("-----------------------");

    ConsoleKeyInfo tecla;
    var text = "";

    do
    {
        // captura somente a tecla
        tecla = Console.ReadKey();

        // verifica se é a tecla enter, se for pula pra outra linha
        // isso tem que ser feito, se nao a linha atual é sobreescrita
        if (tecla.Key == ConsoleKey.Enter)
        {
             Console.WriteLine("");
        }

        // concatena os valores
        text += tecla.KeyChar;
        } while (tecla.Key != ConsoleKey.Escape);

         // remove esc que é um caracter invisivel como mostra na imagem abaixo
         // obs: pode ser só no macOsx isso também
        text = text.Substring(0, text.Length - 1);
        Salvar(text);
}
 ```
 
<img width="296" alt="Screen Shot 2022-05-19 at 22 50 05" src="https://user-images.githubusercontent.com/19297856/169432523-71a9636b-0724-486e-a997-da71e0f67e54.png">

**Reply by andrebaltieri** [reply]
<3
